flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/TwoPhaseCommitSinkFunction.java

```diff
@@ -0,0 +1,342 @@
+
+		preCommit(currentTransaction);
+		pendingCommitTransactions.add(new TransactionAndCheckpoint<>(currentTransaction, checkpointId));
+		LOG.debug("{} - stored pending transactions {}", name(), pendingCommitTransactions);
+
+		currentTransaction = beginTransaction();
+		LOG.debug("{} - started new transaction '{}'", name(), currentTransaction);
+
+		pendingCommitTransactionsState.clear();
+		for (TransactionAndCheckpoint<TXN> pendingCommitTransaction : pendingCommitTransactions) {
+			pendingCommitTransactionsState.add(pendingCommitTransaction);
+		}
+
+		pendingTransactionsState.clear();
+		// in case of failure we might not be able to abort currentTransaction. Let's store it into the state
+		// so it can be aborted after a restart/crash
+		pendingTransactionsState.add(currentTransaction);
+	}
+
+	@Override
+	public final void initializeState(FunctionInitializationContext context) throws Exception {
+		// when we are restoring state with pendingCommitTransactions, we don't really know whether the
+		// transactions were already committed, or whether there was a failure between
+		// completing the checkpoint on the master, and notifying the writer here.
+
+		// (the common case is actually that is was already committed, the window
+		// between the commit on the master and the notification here is very small)
+
+		// it is possible to not have any transactions at all if there was a failure before
+		// the first completed checkpoint, or in case of a scale-out event, where some of the
+		// new task do not have and transactions assigned to check)
+
+		// we can have more than one transaction to check in case of a scale-in event, or
+		// for the reasons discussed in the 'notifyCheckpointComplete()' method.
+
+		pendingTransactionsState = context.getOperatorStateStore().getListState(pendingTransactionsDescriptor);
+		pendingCommitTransactionsState = context.getOperatorStateStore().getListState(pendingCommitTransactionsDescriptor);
+
+		int subtaskIndex = getRuntimeContext().getIndexOfThisSubtask();
+		if (context.isRestored()) {
+			LOG.info("{} - restoring state (subtaskIndex={}).", name(), subtaskIndex);
+
+			for (TransactionAndCheckpoint<TXN> recoveredTransaction : pendingCommitTransactionsState.get()) {
+				recoverAndCommit(recoveredTransaction.transaction);
+				LOG.debug("{} committed recovered transaction {} (subtaskIndex={})", name(), recoveredTransaction, subtaskIndex);
+			}
+
+			// Explicitly abort transactions that could be not closed cleanly
+			for (TXN pendingTransaction : pendingTransactionsState.get()) {
+				recoverAndAbort(pendingTransaction);
+				LOG.debug("{} aborted recovered transaction {} (subtaskIndex={})", name(), pendingTransaction, subtaskIndex);
+			}
+		} else {
+			LOG.info("{} - no state to restore {} (subtaskIndex={}).", name(), subtaskIndex);
+		}
+		this.pendingCommitTransactions.clear();
+	}
+
+	@Override
+	public void open(Configuration parameters) throws Exception {
```

Comment:
In a function that has `initializeState()` you can also put this code in there because `initializeState()` will always be called. Having both `open()` and `initializeState()` is tricky because of the ordering in which those two are called (this has lead to problems in the past when people where changing code and not completely aware of the ordering). For the record, currently the ordering is: 1) `initializeState()` 2) `open()`.